allennlp/modules/encoder_base.py

```diff
@@ -0,0 +1,268 @@
+        # returning the state/s.
+        if len(self._states) == 1:
+            # GRUs only have a single state. This `unpacks` it from the
+            # tuple and returns the tensor directly.
+            correctly_shaped_state = correctly_shaped_states[0]
+            sorted_state = correctly_shaped_state.index_select(1, sorting_indices)
+            return sorted_state[:, :num_valid, :]
+        else:
+            # LSTMs have a state tuple of (state, memory).
+            sorted_states = [state.index_select(1, sorting_indices)
+                             for state in correctly_shaped_states]
+            return tuple(state[:, :num_valid, :] for state in sorted_states)
+
+    def _update_states(self,
+                       final_states: RnnStateStorage,
+                       restoration_indices: torch.LongTensor) -> None:
+        """
+        After the RNN has run forward, the states need to be updated.
+        This method just sets the state to the updated new state, performing
+        several pieces of book-keeping along the way - namely, unsorting the
+        states and ensuring that the states of completely padded sequences are
+        not updated.
+
+        Parameters
+        ----------
+        final_states : ``RnnStateStorage``, required.
+            The hidden states returned as output from the RNN.
+        restoration_indices : ``torch.LongTensor``, required.
+            The indices that invert the sorting used in ``sort_and_run_forward``
+            to order the states with respect to the lengths of the sequences in
+            the batch.
+        """
+        # TODO(Mark): seems weird to sort here, but append zeros in the subclasses.
+        # which way around is best?
+        new_sorted_states = [state.index_select(1, restoration_indices)
+                             for state in final_states]
+
+        if self._states is None:
+            # We don't already have states, so just set the
+            # ones we receive to be the current state.
+            self._states = tuple([torch.autograd.Variable(state.data)
+                                  for state in new_sorted_states])
+        else:
+            # Now we've sorted the states back so that they correspond to the original
+            # indices, we need to figure out what states we need to update, because if
+            # didn't use a state for a particular row, we want to preserve it's state.
+            # Thankfully, the rows which are all zero in the state correspond exactly
+            # to those which aren't used, so we create masks of shape (new_batch_size,),
+            # denoting which states were used in the RNN computation.
+            current_state_batch_size = self._states[0].size(1)
+            new_state_batch_size = final_states[0].size(1)
+            # Masks for the unused states of shape (1, new_batch_size, 1)
+            used_new_rows_mask = [(state[0, :, :].sum(-1) != 0.0).float().view(1, -1, 1)
```

Comment:
Why is this expensive?

Reply:
Because it's a GPU operation, and all operations have overhead.  It's not that this particular operation is slow (like `.data[0]` is), it just seemed like it was unnecessary.  I then realized that it probably isn't.